src/main/scala/lms/core/backend.scala

```diff
@@ -325,15 +325,13 @@ class GraphBuilder {
           ??? // FIXME what about @, ?, array_apply => conservative write on all args?
       }
 
-      // For @ we need to transfer effect on parameters
-      // on the actual values
-      for ((x: Exp, xin) <- args zip argsSym) {
-          if (fwrites(xin)) writes += x
-          else if (freads(xin))  reads += x
-      }
+      // For @ we need to replace the effect on parameters to the actual arguments.
+      // the asInstanceOf seems unsafe at first glance. However, it is not a problem since a standalone block
+      // should never be an argument in function application.
+      val updated_fwrites: Set[Exp] = fwrites.map(k => if (argsSym.contains(k)) args(argsSym.indexOf(k)).asInstanceOf[Exp] else k)
+      val updated_freads: Set[Exp] = freads.map(k => if (argsSym.contains(k)) args(argsSym.indexOf(k)).asInstanceOf[Exp] else k)
+      (reads ++ updated_freads, writes ++ updated_fwrites)
```

Comment:
Here we simply replace the effect symbols for parameters to the corresponding arguments. All other effect symbols remain unchanged and are added to the `reads` and `writes`.